pkg/operators/wasm/datasource.go

```diff
@@ -0,0 +1,486 @@
+// Params:
+// - stack[0]: DataSource handle
+// Return value:
+// - data handle, 0 on error
+func (i *wasmOperatorInstance) dataSourceNewPacketSingle(ctx context.Context, m wapi.Module, stack []uint64) {
+	ds, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.DataSource)
+	if !ok {
+		i.logger.Warnf("DataSource handle %d not found", stack[0])
+		stack[0] = 0
+		return
+	}
+	packet, err := ds.NewPacketSingle()
+	if err != nil {
+		i.logger.Warnf("failed to create NewPacketSingle: %v", err)
+		stack[0] = 0
+		return
+	}
+	stack[0] = wapi.EncodeU32(i.addHandle(packet))
+}
+
+// dataSourceNewPacketArray allocates and returns a handle to a new data instance.
+// Params:
+// - stack[0]: DataSource handle
+// Return value:
+// - data handle, 0 on error
+func (i *wasmOperatorInstance) dataSourceNewPacketArray(ctx context.Context, m wapi.Module, stack []uint64) {
+	ds, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.DataSource)
+	if !ok {
+		i.logger.Warnf("DataSource handle %d not found", stack[0])
+		stack[0] = 0
+		return
+	}
+	packet, err := ds.NewPacketArray()
+	if err != nil {
+		i.logger.Warnf("failed to call NewPacketArray: %v", err)
+		stack[0] = 0
+		return
+	}
+	stack[0] = wapi.EncodeU32(i.addHandle(packet))
+}
+
+// dataSourceEmitAndRelease emits and releases the data.
+// Params:
+// - stack[0]: Data handle
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) dataSourceEmitAndRelease(ctx context.Context, m wapi.Module, stack []uint64) {
+	ds, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.DataSource)
+	if !ok {
+		i.logger.Warnf("DataSource handle %d not found", stack[0])
+		stack[0] = 1
+		return
+	}
+	packet, ok := i.getHandle(wapi.DecodeU32(stack[1])).(datasource.Packet)
+	if !ok {
+		i.gadgetCtx.Logger().Warnf("data handle %d not found", stack[1])
+		stack[0] = 1
+		return
+	}
+	if err := ds.EmitAndRelease(packet); err != nil {
+		stack[0] = 1
+		return
+	}
+	stack[0] = 0
+}
+
+// dataSourceRelease releases the data.
+// Params:
+// - stack[0]: Data handle
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) dataSourceRelease(ctx context.Context, m wapi.Module, stack []uint64) {
+	ds, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.DataSource)
+	if !ok {
+		i.logger.Warnf("DataSource handle %d not found", stack[0])
+		stack[0] = 1
+		return
+	}
+	packet, ok := i.getHandle(wapi.DecodeU32(stack[1])).(datasource.Packet)
+	if !ok {
+		i.gadgetCtx.Logger().Warnf("data handle %d not found", stack[1])
+		stack[0] = 1
+		return
+	}
+	ds.Release(packet)
+	stack[0] = 0
+}
+
+// dataArrayNew allocates and returns a new element on the array
+// Params:
+// - stack[0]: DataArray handle
+// Return value:
+// - Data handle, 0 on error
+func (i *wasmOperatorInstance) dataArrayNew(ctx context.Context, m wapi.Module, stack []uint64) {
+	dataArray, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.DataArray)
```

Comment:
Not just here, but in general, do you think it is worth to distinguish between not finding the error and failing the assert?

Reply:
Implemented a new version of getHandle that covers all of this so we don't need to duplicated this logic all over the place.